In [94]:
import numpy as np

from sklearn import datasets, model_selection, naive_bayes
from pandas import DataFrame

In [95]:
%pylab inline

Populating the interactive namespace from numpy and matplotlib


/usr/lib/python3.6/site-packages/IPython/core/magics/pylab.py:161: UserWarning: pylab import has clobbered these variables: ['clf']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"


In [96]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [97]:
print("breast_cancer: {}".format(breast_cancer.keys()))
print("digits: {}".format(digits.keys()))

breast_cancer: dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])
digits: dict_keys(['data', 'target', 'target_names', 'images', 'DESCR'])


In [98]:
digits_frame = DataFrame(digits.data)
digits_frame['target'] = digits.target

In [99]:
digits_frame.head()

0    1    2     3     4     5    6    7    8    9   ...     55   56   57  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0   ...    0.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   

    58    59    60    61   62   63  target  
0  6.0  13.0  10.0   0.0  0.0  0.0       0  
1  0.0  11.0  16.0  10.0  0.0  0.0       1  
2  0.0   3.0  11.0  16.0  9.0  0.0       2  
3  7.0  13.0  13.0   9.0  0.0  0.0       3  
4  0.0   2.0  16.0   4.0  0.0  0.0       4  

[5 rows x 65 columns]

In [100]:
breast_cancer_frame = DataFrame(breast_cancer.data)
breast_cancer_frame.columns = breast_cancer.feature_names
breast_cancer_frame['target'] = breast_cancer.target

In [101]:
breast_cancer_frame.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension   ...    worst texture  worst perimeter  worst area  \
0                 0.07871   ...            17.33           184.60      2019.0   
1                 0.05667   ...            23.41           158.80      1956.0   
2                 0.05999   ...            25.53           152.50      1709.0   
3                 0.09744   ...            26.50            98.87       567.7   
4                 0.05883   ...            16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

По первым нескольким строкам выборки можно заметить, что в датасете digits признаки имеют целочисленные значения, а в датасете breast_canser - вещественные.

In [102]:
clf = naive_bayes.BernoulliNB()

arr_digit_bern = model_selection.cross_val_score(clf, digits.data, digits.target)
arr_cancer_bern = model_selection.cross_val_score(clf, breast_cancer.data, breast_cancer.target)

digit_mean_bern = mean(arr_digit_bern)
cancer_mean_bern = mean(arr_cancer_bern)
print("digit: {}\nbreast_cancer: {}".format(digit_mean_bern, cancer_mean_bern))

digit: 0.8258236507780582
breast_cancer: 0.6274204028589994


In [103]:
clf = naive_bayes.MultinomialNB()

arr_digit_mult = model_selection.cross_val_score(clf, digits.data, digits.target)
arr_cancer_mult = model_selection.cross_val_score(clf, breast_cancer.data, breast_cancer.target)

digit_mean_mult = mean(arr_digit_mult)
cancer_mean_mult = mean(arr_cancer_mult)
print("digit: {}\nbreast_cancer: {}".format(digit_mean_mult, cancer_mean_mult))

digit: 0.8708771489735053
breast_cancer: 0.8945790401930752


In [104]:
clf = naive_bayes.GaussianNB()

arr_digit_gaus = model_selection.cross_val_score(clf, digits.data, digits.target)
arr_cancer_gaus = model_selection.cross_val_score(clf, breast_cancer.data, breast_cancer.target)

digit_mean_gaus = mean(arr_digit_gaus)
cancer_mean_gaus = mean(arr_cancer_gaus)
print("digit: {}\nbreast_cancer: {}".format(digit_mean_gaus, cancer_mean_gaus))

digit: 0.8186003803550138
breast_cancer: 0.9367492806089297


Максимальное качество классификации на датасете digits получилось при использовании MultinomialNB, оно равно 0.87.
Максимальное качество классификации на датасете breast_cancer получилось при использовании GaussianNB, оно равно 0.94.
Полученные результаты согласуются с ожиданиями, так как на вещественной выборке лучше работает нормальное распределение, а на целочисленной - полиномиальное. Распределение Бернулли же работает хуже, чем нормальное и полиномиальное.
Верными утверждениями являются (c) и (d).